hogben/optimise.py

```diff
@@ -156,16 +214,80 @@ def _non_decreasing(x):
         )
         return res[:num_contrasts], res[num_contrasts:], val
 
+    def optimise_parameters(
+            self,
+            angle_times,
+            inst_or_path='OFFSPEC',
+            workers=-1,
+            verbose=True,
+    ) -> tuple:
+        """
+        Finds the optimal parameters for a given sample.
+
+        Args:
+            angle_times (list): points and times for each angle to simulate.
+            inst_or_path: either the name of an instrument already in HOGBEN,
+                          or the path to a direct beam file, defaults to
+                          'OFFSPEC'
+            workers (int): number of CPU cores to use when optimising.
+            verbose (bool): whether to display progress or not.
+
+        Returns:
+            tuple: optimised underlayer parameters and the corresponding
+                   optimisation function value.
+
+        """
+        # Check that the underlayers of the sample can be varied.
+        bounds = []
+        params = self.sample.get_param_by_attribute('optimize')
+        for parameter in params:
+            bounds += [(parameter.bounds.lb, parameter.bounds.ub)]
+        # Arguments for the optimisation function.
+        args = [params, angle_times, inst_or_path]
+
+        # Optimise parameters and return the results.
+        res, val = Optimiser.__optimise(
+            self._parameter_func, bounds, [], args, workers, verbose
+        )
+        return res, val
+
+    def _parameter_func(self,
+                        x: list,
+                        params,
+                        angle_times: list,
+                        inst_or_path: str = 'OFFSPEC') -> float:
+        """Defines the function for optimising arbitrary parameters in sample.
+
+        Args:
+            x (list): parameter values to calculate with.
+            angle_times (type): points and times for each angle.
+            contrasts (list): contrasts of the experiment, if applicable.
+            inst_or_path: either the name of an instrument already in HOGBEN,
+                          or the path to a direct beam file, defaults to
+                          'OFFSPEC'
+
+        Returns:
+            float: negative of minimum eigenvalue using given conditions.
```

Comment:
negative of minimum eigenvalue using given conditions. -->negative of minimum eigenvalue of the Fisher information matrix using the given conditions.

Reply:
In case you're not aware by the way, nowadays you can also add suggestions in GitHub, so that these can just be accepted by the person who submitted the PR with a press of a button like this:
``` 
```suggestion
            float: negative of minimum eigenvalue using given conditions.
\```
```
Without the backslash on the last ``` here. This results on something like this:

```suggestion
            float: negative of minimum eigenvalue using given conditions.
```